File: ios/App/Sources/Status/UseSmileIDSampleStatusApi.swift
import Foundation
import SampleUI

/// `GET /v3/status/{jobId}`, the partner's own call, the SDK stopping at the 202; `URLSession` stays in the shell.
struct UseSmileIDSampleStatusApi: UseSmileIDSampleJobStatusSource {
  private let session: URLSession

  init(session: URLSession = UseSmileIDSampleStatusApi.bounded) {
    self.session = session
  }

  /// Ten seconds, the same on every platform: `URLSession.shared` would wait a minute.
  private static let bounded: URLSession = {
    let configuration = URLSessionConfiguration.default
    configuration.timeoutIntervalForRequest = 10
    // Named rather than left to the loupe's swizzle: this session is built once, and whether that
    // happens before or after the loupe installs is an ordering nobody should have to reason about
    #if DEBUG
      Loupe.instrument(configuration)
    #endif
    return URLSession(configuration: configuration)
  }()

  func check(jobId: String, token: String, sandbox: Bool) async throws -> UseSmileIDSampleStatusRefresh {
    guard let url = useSmileIDSampleStatusUrl(jobId: jobId, sandbox: sandbox) else {
      throw URLError(.badURL)
    }
    var request = URLRequest(url: url)
    // The session's own JWT from `POST /v3/token`. Never logged.
    request.setValue(token, forHTTPHeaderField: "SmileID-Token")
    let (data, response) = try await session.data(for: request)
    return useSmileIDSampleStatusOutcome(
      code: (response as? HTTPURLResponse)?.statusCode ?? 0,
      // Unknown keys are ignored, so a field added server-side cannot turn a good response into a failure.
      body: try? JSONDecoder().decode(UseSmileIDSampleStatusResponse.self, from: data)
    )
  }
}

/// Percent-encoded as one path segment, or a `/`, `?` or `#` would change which request the token is sent with.
func useSmileIDSampleStatusUrl(jobId: String, sandbox: Bool) -> URL? {
  let environment: UseSmileIDSampleEnvironment = sandbox ? .sandbox : .production
  let unreserved = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-._~"))
  guard let segment = jobId.addingPercentEncoding(withAllowedCharacters: unreserved), !segment.isEmpty else {
    return nil
  }
  return URL(string: environment.baseUrl + "v3/status/\(segment)")
}

/// The HTTP code and body onto an outcome. Pure, so the branch table is unit-testable.
func useSmileIDSampleStatusOutcome(
  code: Int,
  body: UseSmileIDSampleStatusResponse?
) -> UseSmileIDSampleStatusRefresh {
  guard let body, (200..<300).contains(code) else { return .failed(reason: "HTTP \(code)") }
  if body.status == UseSmileIDSampleStatusResponse.processing {
    return .stillProcessing
  }
  guard let status = body.sampleStatus else {
    return .failed(reason: "Unrecognised status '\(body.status)'")
  }
  return .updated(status: status, message: body.message, httpCode: code)
}

struct UseSmileIDSampleStatusResponse: Decodable {
  let status: String
  let message: String
  /// Carried for the shape, not read: required, a 2xx omitting one would decode to nil.
  let jobId: String?
  let userId: String?
  let createdAt: String?

  /// Five API states onto the four badges: `error` lands on Blocked and leans on the server's message.
  var sampleStatus: UseSmileIDSampleStatus? {
    switch status {
    case "clear": .clear
    case "attention": .attention
    case "block", "error": .blocked
    default: nil
    }
  }

  static let processing = "processing"

  enum CodingKeys: String, CodingKey {
    case status
    case message
    case jobId = "job_id"
    case userId = "user_id"
    case createdAt = "created_at"
  }
}

File: expo/app/src/status/use-smile-id-sample-status-api.ts
import {
  UseSmileIDSampleStatus,
  smileIDSampleEnvironmentBaseUrl,
  type UseSmileIDSampleJobStatusSource,
  type UseSmileIDSampleStatusRefresh,
} from '@smileid/sample-ui';

/// Ten seconds, the same on every platform.
const TIMEOUT_MILLIS = 10_000;

/// One path segment, or a `/`, `?` or `#` would change the request.
export const smileIDSampleStatusUrl = (jobId: string, sandbox: boolean): string | null => {
  const segment = encodeURIComponent(jobId);
  if (segment.length === 0) return null;
  return `${smileIDSampleEnvironmentBaseUrl(sandbox ? 'sandbox' : 'production')}v3/status/${segment}`;
};

/// The HTTP code and body onto an outcome.
export const smileIDSampleStatusOutcome = (code: number, body: unknown): UseSmileIDSampleStatusRefresh => {
  const response = body as { status?: unknown; message?: unknown } | null;
  if (response === null || typeof response !== 'object' || code < 200 || code >= 300) {
    return { kind: 'failed', reason: `HTTP ${code}` };
  }
  if (typeof response.status !== 'string' || typeof response.message !== 'string') {
    return { kind: 'failed', reason: `HTTP ${code}` };
  }
  if (response.status === 'processing') return { kind: 'stillProcessing' };
  const status = statusFor(response.status);
  if (status === null) return { kind: 'failed', reason: `Unrecognised status '${response.status}'` };
  return { kind: 'updated', status, message: response.message, httpCode: code };
};

/// Five API states onto four badges; `error` is Blocked.
const statusFor = (status: string): UseSmileIDSampleStatus | null => {
  switch (status) {
    case 'clear':
      return UseSmileIDSampleStatus.Clear;
    case 'attention':
      return UseSmileIDSampleStatus.Attention;
    case 'block':
    case 'error':
      return UseSmileIDSampleStatus.Blocked;
    default:
      return null;
  }
};

/// `GET /v3/status/{jobId}`, the partner's own call.
export const smileIDSampleStatusApi: UseSmileIDSampleJobStatusSource = {
  check: async (jobId, token, sandbox) => {
    const url = smileIDSampleStatusUrl(jobId, sandbox);
    if (url === null) throw new TypeError('No job id to ask about');
    const abort = new AbortController();
    const timer = setTimeout(() => abort.abort(), TIMEOUT_MILLIS);
    try {
      // The session's JWT; never logged.
      const response = await fetch(url, { headers: { 'SmileID-Token': token }, signal: abort.signal });
      const body: unknown = await response.json().catch(() => null);
      return smileIDSampleStatusOutcome(response.status, body);
    } finally {
      clearTimeout(timer);
    }
  },
};
